File: code_frac.py
import matplotlib.pyplot as plt
import numpy as np
import argparse

# Parse command-line arguments
parser = argparse.ArgumentParser(description="Générer une fractale et l'enregistrer dans un fichier BMP.")
parser.add_argument("--taille", type=int, default=1000, help="Taille de l'image en pixels (par défaut 10000)")
parser.add_argument("--iterations", type=int, default=1000, help="Nombre maximal d'itérations (par défaut 1000)")
parser.add_argument("--output", type=str, default="fractale.bmp", help="Nom du fichier de sortie (par défaut fractale.bmp)")
args = parser.parse_args()

# Paramètres
xmin, xmax = -1, 1
ymin, ymax = -1, 1
a = -0.8
b = 0.156
taille = args.taille
iterationmax = args.iterations

# Initialisation des pixels
pixels = np.zeros((taille, taille, 3), dtype='uint8')

# Génération de la fractale
for line in range(taille):
    for col in range(taille):
        i = 1
        x = xmin + col * (xmax - xmin) / taille
        y = ymax - line * (ymax - ymin) / taille
        while i <= iterationmax and (x**2 + y**2) <= 4:
            x, y = x**2 - y**2 + a, 2 * x * y + b
            i += 1
        if i > iterationmax and (x**2 + y**2) <= 4:
            pixels[col, line] = (0, 0, 0)
        else:
            pixels[col, line] = ((4 * i) % 256, 2 * i % 256, (6 * i) % 256)

# Sauvegarde de l'image dans un fichier BMP
plt.imsave(args.output, pixels)

print(f"Fractale générée et enregistrée dans le fichier {args.output}")

File: code_frac.c
#include "libbmp/libbmp.h"
#include <stdio.h>
#include <stdlib.h>

void generateFractal(const char *filename, int taille, int iterationmax) {
    double xmin = -1.0, xmax = 1.0;
    double ymin = -1.0, ymax = 1.0;
    double a = -0.8, b = 0.156;

    // Création de l'image BMP
    bmp_img img;
    bmp_img_init_df(&img, taille, taille);

    // Génération de la fractale
    for (int y = 0; y < taille; y++) {
        for (int x = 0; x < taille; x++) {
            int i = 1;
            double cx = xmin + x * (xmax - xmin) / taille;
            double cy = ymax - y * (ymax - ymin) / taille;
            double zx = cx, zy = cy;

            while (i <= iterationmax && (zx * zx + zy * zy) <= 4) {
                double xtemp = zx * zx - zy * zy + a;
                zy = 2 * zx * zy + b;
                zx = xtemp;
                i++;
            }

            if (i > iterationmax && (zx * zx + zy * zy) <= 4) {
                bmp_pixel_init(&img.img_pixels[y][x], 0, 0, 0); // Pixel noir pour les points convergents
            } else {
                bmp_pixel_init(&img.img_pixels[y][x], (4 * i) % 256, (2 * i) % 256, (6 * i) % 256);
            }
        }
    }

    // Sauvegarde de l'image
    if (bmp_img_write(&img, filename) != BMP_OK) {
        fprintf(stderr, "Erreur lors de la sauvegarde du fichier BMP.\n");
    } else {
        printf("Fractale générée et sauvegardée dans le fichier %s\n", filename);
    }

    bmp_img_free(&img);
}

int main(int argc, char *argv[]) {
    int taille = 10000;       // Par défaut : 1000x1000 pixels
    int iterations = 1000;   // Par défaut : 1000 itérations
    const char *output = "fractale.bmp";

    if (argc > 1) taille = atoi(argv[1]);
    if (argc > 2) iterations = atoi(argv[2]);
    if (argc > 3) output = argv[3];

    generateFractal(output, taille, iterations);
    return 0;
}
